p2p/tools/factories/peer.py

```
@@ -0,0 +1,134 @@
+                          bob_p2p_version: int = 5,
+                          cancel_token: CancelToken = None,
+                          event_bus: EndpointAPI = None,
+                          ) -> AsyncIterator[Tuple[BasePeer, BasePeer]]:
+    if cancel_token is None:
+        cancel_token = CancelTokenFactory()
+
+    alice_factory = alice_peer_factory_class(
+        privkey=alice_private_key,
+        context=alice_peer_context,
+        token=cancel_token,
+        event_bus=event_bus,
+    )
+    bob_factory = bob_peer_factory_class(
+        privkey=bob_private_key,
+        context=bob_peer_context,
+        token=cancel_token,
+        event_bus=event_bus,
+    )
+
+    # start: ConnectionPairFactory
+
+    alice_transport, bob_transport = MemoryTransportPairFactory(
+        alice_remote=alice_remote,
+        alice_private_key=alice_private_key,
+        bob_remote=bob_remote,
+        bob_private_key=bob_private_key,
+    )
+
+    # end: ConnectionPairFactory
+
+    alice = alice_factory.create_peer(transport=alice_transport, inbound=False)
+    bob = bob_factory.create_peer(transport=bob_transport, inbound=False)
+
+    async def do_handshake(peer: BasePeer) -> None:
+        await peer.do_p2p_handshake()
+        await peer.do_sub_proto_handshake()
+
+    await asyncio.wait_for(asyncio.gather(
+        do_handshake(alice),
+        do_handshake(bob),
+    ), timeout=1)
+
+    alice_task = asyncio.ensure_future(alice.run())
+    bob_task = asyncio.ensure_future(bob.run())
+
+    await asyncio.wait_for(asyncio.gather(
+        alice.events.started.wait(),
+        bob.events.started.wait(),
+    ), timeout=1)
+
+    try:
+        yield alice, bob
+    finally:
+        await asyncio.wait_for(asyncio.gather(
+            alice.cancel(),
+            bob.cancel(),
+        ), timeout=1)
+        alice_task.cancel()
+        bob_task.cancel()
+        try:
+            await alice_task
+        except asyncio.CancelledError:
```

Is this just to silence errors about uncaught `CancelledError`s? Seems like it's otherwise not import to await these tasks. Either way, would be nice to have a little comment here.

---

This is the *proper* way to cancel something that was scheduled with `ensure_future`.  You first cancel it, then you await it, catching `CancelledError`.  (this approach is from the `asyncio` docs)

---

Hah, well even the basics of `asyncio` still surprise me some days. Carry on, nothing to see here...